src/models/tools/shared-model.ts

```diff
@@ -0,0 +1,128 @@
+// If UnknownContentModel is moved to its own module this circular dependency causes an error.
+// If they are in the same module then this isn't a problem.
+//
+// There is a still an "unknown-content" module, so that module can
+// register the tool without adding a circular dependency on tool-content-info here.
+const _UnknownSharedModel = SharedModel
+  .named("UnknownSharedModel")
+  .props({
+    type: types.optional(types.literal(kUnknownSharedModel), kUnknownSharedModel),
+    original: types.maybe(types.string)
+  });
+
+export const UnknownSharedModel = types.snapshotProcessor(_UnknownSharedModel, {
+  // Maybe we can type the snapshot better?
+  preProcessor(snapshot: any) {
+    const type = snapshot?.type;
+    return type && (type !== kUnknownSharedModel)
+            ? {
+              type: kUnknownSharedModel,
+              original: JSON.stringify(snapshot)
+            }
+            : snapshot;
+  },
+
+  postProcessor(snapshot: any) {
+    return JSON.parse(snapshot.original);
+  }
+});
+
+/**
+ * An instance of this interface should be provided to tiles so they can interact
+ * with shared models.
+ */
+export interface ISharedModelManager {
+  /**
+   * The manager might be available, but is not ready to be used yet.
+   */
+  get isReady(): boolean;
+  
+  /**
+   * Find the shared model at the container level. If the tile wants to use this
+   * shared model it should call `setTileSharedModel` to save a reference to it.
+   * The container needs to know which tiles reference which shared models so it
+   * can update them when the shared model changes.
+   *
+   * @param sharedModelType the MST model "class" of the shared model
+   */
+  findFirstSharedModelByType<IT extends typeof SharedModelUnion>(sharedModelType: IT): IT["Type"] | undefined;
+
+  /**
+   * Get the shared model of this tile. They are labeled so a tile can have
+   * multiple shared models.
+   *
+   * @param tileContentModel normally this would be `self` when called by a tile
+   * @param label a string labeling this shared model
+   */ 
+  getTileSharedModel(tileContentModel: IAnyStateTreeNode, label: string): SharedModelType | undefined;
+
+  /**
+   * Tiles should call this after finding or creating shared model they want to
+   * use. If this is a new shared model instance, it will be added to the
+   * container level so other tiles can find it.
+   *
+   * It is important that tiles call this even if they find an existing shared
+   * model with `findFirstSharedModelByType`. The container needs to know which
+   * tiles are referencing which shared models.
+   *
+   * @param tileContentModel normally this would be `self` when called by a tile
+   * @param label a string labeling this shared model
+   * @param sharedModel a new shared model instance or one returned by
+   * `findFirstSharedModelByType`
+   */
+  setTileSharedModel(tileContentModel: IAnyStateTreeNode, label: string, sharedModel: SharedModelType): void;
```

Comment:
Yes I can see how this name is confusing. 

The idea of the label/key was so the code of a tile can refer to multiple shared models using a label/key to identify which is which.  Perhaps a tile wants to work with a variables shared model and a data set shared model. So it would use keys of "variables" and "dataSet".  And if it wanted to work with multiple data sets it could do that with "dataSet1", "dataSet2".

I was thinking that this would be the only reference to the shared model, so it would be necessary to have a way for a tile that needs to differentiate between two models of the same type to give them labels, or keys. This approach means there is only one place that needs to be updated if we have to change the sharedModel id. So basically any reference to the shared model is indirect through the label or key. 

However the tile will have references to items in the shared model, so we have to go into the tile state anyway to find an update those ids. 

Based on that I'm now thinking I can just drop the label/key. And the functions can just be `getSharedModelsOfTile(tile)`, and `addSharedModelToTile(tile, sharedModel)`. 

If a tile needs to differentiate between multiple shared models it can save references to the shared model in its own content. This would mean a tile could store an ordered array of dataSets instead of having to manage keys like "dataSet1", "dataSet2", ...

The only danger I see is that a developer might forget to call `addSharedModelToTile` and just store a reference in its content. But hopefully the developer would quickly find out that the the `updateAfterSharedModelChanges` was not being called in this case.

Reply:
This is addressed in this PR: https://github.com/concord-consortium/collaborative-learning/pull/1256